### ISSUE TITLE
GaussianBlurEffect (eg. in AcrylicBrush) doesn't work over a SwapChainPanel

### ISSUE THREAD

**Issue by Sergio0694** [microsoft/microsoft-ui-xaml#4947]
**Describe the bug**
It looks like the Win2D [`GaussianBlurEffect`](https://microsoft.github.io/Win2D/WinUI2/html/T_Microsoft_Graphics_Canvas_Effects_GaussianBlurEffect.htm) doesn't work when applied on a brush/surface that's above a [`SwapChainPanel`](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.swapchainpanel). In my case I'm using an [`IDXGISwapChain2`](https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_3/nn-dxgi1_3-idxgiswapchain2) with DX12 APIs. The interesting thing is that it's not like the swap chain is just treated as "on top", with nothing else going over it. Instead, you can place UI elements over it just fine, and all other Composition/Win2D effects seem to work/sample just fine over it as well, it's just the blur that's not working. For instance, the various tint/exposure/border/etc. effects correctly overlay on top of the swap chain. Is this a bug with just the blur effect?

> **NOTE:** I can repro this with both the blur effect from `AcrylicBrush`, as well as with a custom effects pipeline. In both cases, blur just won't be applied over the swap chain content, unlike all other effects in the pipeline in use.

**Steps to reproduce the bug**

Admittedly this bug requires a few steps to properly repro, due to the number of components involved.
Here's the most minimal repro I can provide:
1. Clone [`ComputeSharp`](https://github.com/Sergio0694/ComputeSharp/)
2. Checkout to the `feature/winui3-shader-panel` branch
3. Create a new WinUI 3 Desktop app
4. Add a reference to the `ComputeSharp.WinUI` project, and the `ComputeSharp.SourceGenerators` as an analyzer
4. Use the following XAML in `MainWindow`:
```xml
<Grid>
    <Grid.Background>
        <ImageBrush ImageSource="http://3.bp.blogspot.com/-84ZrGEHfHr0/UJtqvO07w-I/AAAAAAAAAv0/dMeI3WM5wWA/s1600/Nature-HD-Wallpaper1.jpg"/>
    </Grid.Background>
    <computesharp:ComputeShaderPanel
        x:Name="ShaderPanel"
        Height="400"
        Width="560"/>
    <Border>
        <Border.Background>
            <AcrylicBrush
                BackgroundSource="Backdrop"
                TintColor="Gray"
                TintOpacity="0.2"/>
        </Border.Background>
    </Border>
</Grid>
```
5. Add any of the shaders from `samples/ComputeSharp.SwapChain/Shaders` to the project
6. In `MainWindow.xaml.cs`, add this to the constructor (changing the shader type if needed):
```csharp
ShaderPanel.ShaderRunner = new ShaderRunner<TwoTiledTruchet>(static (texture, time) => new(texture, (float)time.TotalSeconds));
```

**Expected behavior**

The `AcrylicBrush` in use should correctly be applied on the entire window content.
Instead, the blur won't work just over the swap chain (but it will in the rest of the window).
The other acrylic effects will work fine over the swap chain (eg. noise, tint, etc.).

**Screenshots**

![mgImyNt6Bn](https://user-images.githubusercontent.com/10199417/116914325-e4232d00-ac4a-11eb-8905-eb072098b224.gif)

**Version Info**

- `Microsoft.ProjectReunion` version `0.5.6`
- `Microsoft.ProjectReunion.Foundation` version `0.5.6`
- `Microsoft.ProjectReunion.WinUI` version `0.5.6`
- `Microsoft.Windows.SDK.NET.Ref` version `10.0.18362.16`

Windows app type:
| UWP              | Win32 (WinUI 3 Desktop)            |
| :--------------- | :--------------- |
|  | Yes |

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| October 2020 Update (19042)        | Yes   |
| May 2020 Update (19041)            |    |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            | Yes |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Comment by StephenLPeters** [comment]
@codendone and @jeffstall I'm not sure which team this issue would be associated with, thoughts?

**Comment by codendone** [comment]
This is a limitation of SwapChainPanel in WinUI 3. The swapchain is given to the Windows compositor to draw below the rest of the WinUI content composed by the in-process WinUI 3 compositor. Any effect which uses other WinUI rendering as input can only use pixels drawn by the in-proc compositor, which does not include the swapchain.

We should fix the documentation to state this limitation.

**Comment by marb2000** [comment]
Perhaps this should be a feature request?

**Comment by Sergio0694** [comment]
Thanks for the additional info @codendone! I figured it was probably something related to WinUI 3 using its own compositor in each window, that makes sense. If I may ask one follow up question - how are the other effects still working even on top of the swap chain though? As in, if I apply an `AcrylicBrush` (or some custom effects pipeline similar to it) over the window, I can see all other effects other than the blur being applied on top of the swap chain. This means that the in-app backdrop brush (which is the first node in the pipeline) is in fact sampling correctly including what's being displayed in the swap chain. That sampling is then passed through the rest of the pipeline where other stuff is eventually added to it (eg. a noise texture, or some tint, etc.), and the resulting brush is applied to a visual that's covering the window. I guess I'm confused as to how can these other effects still work in this scenario - if the swap chain is excluded by what's available to these effects, how can the in-app backdrop brush still "see" it? And if other effects in the pipeline (eg. the blur) are just being applied to previous nodes in the pipeline (eg. the same in-app backdrop brush), how is this input then changed so that the swap chain ends up not being visible to them?

In other words, if my composition brush is covering the whole window and the swap chain cannot be seen at all by other effects (eg. the in-app backdrop brush at the start of the pipeline), then why isn't the area where the swap chain is just black, or simply transparent? Just trying to understand this better, is all (and really appreciate your time!) 🙂

@marb2000 Sounds good to me, it would indeed be very nice to have! And also be more intuitive for developers. Should I just change the title of this issue and the first post, or should I just open a separate issue with the proposal and link this from there?

**Comment by codendone** [comment]
The CompositionBackdropBrush in the AcrylicBrush will sample from what the in-proc compositor has drawn up to that point, which will be transparent pixels in the area where the SwapChainPanel. The rest of the AcrylicBrush recipe will then run with that input and draw final pixels.

Notes:
- The in-proc compositor erases to transparent the area of the SwapChainPanel when it would have otherwise drawn the SwapChainPanel. This allows the swapchain of the SwapChainPanel, which is given to the system Compositor, to be seen below what the in-proc compositor drew.
- If some other non-SwapChainPanel elements are z-order on top of the SwapChainPanel and below the element with the AcrylicBrush, then those elements will still be available for the in-proc compositor to sample from. 

In the case of the AcrylicBrush recipe, it uses the BackdropBrush which (assuming no elements in between) will be fully transparent in the SwapChainPanel area, then it will blur those transparent pixels (which will still be transparent), then the noise gets added which makes the pixels slightly opaque, and finally the tint and luminosity effects will tint and brighten those mostly-transparent pixels to produce a slightly visible effect. I think without the noise layer it would be completely transparent. The blur, of course, does affect the edges of the SwapChainPanel, since it will blend some visible pixels outside the SwapChainPanel into the edges of the SwapChainPanel.

**Comment by Sergio0694** [comment]
Oh I see, that makes sense. Thank you so much for your time and the detailed explanation @codendone! 😄

@marb2000 I'm going to close this issue since it was a question and that's been addressed now, I will open a separate one as a proposal and will link this issue from there for future reference as well. This sure would be a neat feature to have in WinUI 3! 🚀
